diff --git a/lonestar/pagerank/PageRank-pull.cpp b/lonestar/pagerank/PageRank-pull.cpp
index b2eba0c..fb915bf 100644
--- a/lonestar/pagerank/PageRank-pull.cpp
+++ b/lonestar/pagerank/PageRank-pull.cpp
@@ -25,6 +25,7 @@
 #include "galois/graphs/LCGraph.h"
 #include "galois/graphs/TypeTraits.h"
 #include "galois/gstl.h"
+#include <fstream>
 
 const char* desc =
     "Computes page ranks a la Page and Brin. This is a pull-style algorithm.";
@@ -249,6 +250,14 @@ void prResidual(Graph& graph) {
   prTimer.start();
   computePRResidual(graph, delta, residual);
   prTimer.stop();
+
+  float measure_time = prTimer.get()==0?prTimer.get_usec()/float(1000):prTimer.get();
+  measure_time /= float(1000);
+
+  std::ofstream perf_fp;
+  perf_fp.open("perf_result");
+  perf_fp << "PR," << filename << "," << measure_time << ",\n" << std::endl;
+  perf_fp.close();
 }
 
 int main(int argc, char** argv) {
